### PR TITLE
Feature/party system

### PR DESCRIPTION
**Pull request by samuelWilliams99**
image

### CODE REVIEW

File: gamemode/modules/hud/cl_hud.lua [hunk 1]
```diff
@@ -1,6 +1,8 @@
 local leftBg = Material( "rvr/backgrounds/hud_background.png" )
 local rightBg = Material( "rvr/backgrounds/hud_background_main.png" )
 
+local targetIDRange = 100
```
Comment (samuelWilliams99) [inline gamemode/modules/hud/cl_hud.lua]: Move to HUD config

File: gamemode/modules/hud/cl_hud.lua [hunk 2]
```diff
@@ -77,3 +83,37 @@ function GM:HUDShouldDraw( name )
...
     return self.BaseClass.HUDShouldDraw( self, name )
 end
+
+function GM:HUDDrawTargetID()
+    local trace = LocalPlayer():GetEyeTrace()
+    local aimEnt = trace.Entity
+
+    if type( aimEnt ) ~= "Player" then return end
+
+    if LocalPlayer():GetShootPos():Distance( trace.HitPos ) > targetIDRange then return end
```
Comment (samuelWilliams99) [inline gamemode/modules/hud/cl_hud.lua]: Use DistToSqr

File: gamemode/modules/party/cl_party.lua [hunk 3]
```diff
@@ -0,0 +1,149 @@
...
+    if #friends == 0 then return end
+
+    net.Start( "RVR_Party_updateFriends" )
+        net.WriteTable( friends )
+    net.SendToServer()
+end
+
+hook.Add( "InitPostEntity", "RVR_Party_sendFriendData", sendFriendData )
+net.Receive( "RVR_Party_updateFriends", sendFriendData )
+timer.Create( "RVR_Party_updateFriends", 300, 0, sendFriendData ) -- TODO: discuss if this is good?
```
Comment (samuelWilliams99) [inline gamemode/modules/party/cl_party.lua]: check every time a player tries to join

File: gamemode/modules/party/cl_party.lua [hunk 4]
```diff
@@ -0,0 +1,149 @@
...
+
+hook.Add( "PreDrawHalos", "RVR_Party", function()
+    local ownParty = LocalPlayer():GetParty()
+
+    if not ownParty then return end
+    halo.Add( ownParty.members, ownParty.color, 5, 5, 2 )
+end )
+
+local function drawNameLabel( ply, textColor )
+    local me = LocalPlayer()
```
Comment (samuelWilliams99) [inline gamemode/modules/party/cl_party.lua]:
```suggestion
    local localPly = LocalPlayer()
```

Comment (samuelWilliams99) [reply gamemode/modules/party/cl_party.lua]: Consider using target + ply, over ply + localPly

File: gamemode/modules/party/cl_party.lua [hunk 5]
```diff
@@ -0,0 +1,149 @@
...
+
+    local scale = math.Clamp( dist * 0.002, 0.5, 3 )
+
+    local pos = ply:GetShootPos() + Vector( 0, 0, 5 + scale * 15 )
+    local text = ply:Nick()
+
+    cam.Start3D2D( pos, textAng, scale )
+        surface.SetFont( "ChatFont" )
+        local tw, th = surface.GetTextSize( text )
+        draw.SimpleText( text, "ChatFont", -tw / 2, -th / 2, textColor ) -- try TEXT_ALIGN_CENTER
```
Comment (samuelWilliams99) [inline gamemode/modules/party/cl_party.lua]: Do this

File: gamemode/modules/party/cl_party.lua [hunk 6]
```diff
@@ -0,0 +1,149 @@
...
+    if not ownParty then return end
+    for _, ply in pairs( ownParty.members ) do
+        if IsValid( ply ) and ply ~= LocalPlayer() then
+            drawNameLabel( ply, ownParty.color )
+        end
+    end
+end )
+
+-- Allow other addons to prevent this, to do it themself
+if hook.Run( "RVR_ShouldOverrideAddText" ) ~= false then
```
Comment (samuelWilliams99) [inline gamemode/modules/party/cl_party.lua]: Make this a config value, use a hook on config loaded

File: gamemode/modules/party/cl_party_menu.lua [hunk 7]
```diff
@@ -0,0 +1,190 @@
...
+    superContainer.Paint = nil
+    superContainer.prog = party.menuEnabled and 1 or 0
+    function superContainer:Think()
+        if self.prog < 1 and party.menuEnabled then
+            self.prog = math.Clamp( self.prog + FrameTime() * 5, 0, 1 )
+        elseif self.prog > 0 and not party.menuEnabled then
+            self.prog = math.Clamp( self.prog - FrameTime() * 5, 0, 1 )
+        end
+
+
```
Comment (samuelWilliams99) [inline gamemode/modules/party/cl_party_menu.lua]: ew

File: gamemode/modules/party/sv_party.lua [hunk 8]
```diff
@@ -0,0 +1,322 @@
+RVR.Party = RVR.Party or {}
+local party = RVR.Party
+party.idCounter = party.idCounter or 0
+
+-- TODO: Start player as spectator
```
Comment (samuelWilliams99) [inline gamemode/modules/party/sv_party.lua]: spectate why

File: gamemode/modules/party/sv_party.lua [hunk 9]
```diff
@@ -0,0 +1,322 @@
...
+
+    if owner:GetPartyID() then
+        return nil, "Player " .. owner:Nick() .. " already in a party"
+    end
+
+    if #tag ~= 4 then
+        return nil, "Tag must be 4 characters"
+    end
+
+    if #partyName < 5 or tonumber( partyName ) then
```
Comment (samuelWilliams99) [inline gamemode/modules/party/sv_party.lua]: Pull out min and max party name length into config, check for whitespace only names and/or just trim it

File: gamemode/modules/party/sv_party.lua [hunk 10]
```diff
@@ -0,0 +1,322 @@
...
+
+    if mode < 0 or mode > 2 then
+        return false, "Invalid join mode"
+    end
+
+    partyData.joinMode = mode
+    party.broadcastMessage( partyData.id, "Party join mode has been set to " .. party.joinModeStrs[mode] )
+end
+
+-- TODO: Drop inventory, switch to spectator
```
Comment (samuelWilliams99) [inline gamemode/modules/party/sv_party.lua]: Discuss more

File: gamemode/modules/party/sv_party.lua [hunk 11]
```diff
@@ -0,0 +1,322 @@
...
+        party.broadcastMessage( partyData.id, "The owner ( " .. ply:Nick() .. " has left this party, " .. partyData.owner:Nick() .. " is the new owner." )
+    end
+
+    updateClientPartyData( id )
+
+    return true
+end
+
+-- TODO: Teleport to raft
+function party.addMember( id, ply )
```
Comment (samuelWilliams99) [inline gamemode/modules/party/sv_party.lua:153]: Sets spawn point instead, discuss this idea

File: gamemode/modules/party/sv_party.lua [hunk 12]
```diff
@@ -0,0 +1,322 @@
...
+    if table.HasValue( partyData.members, ply ) then
+        return false, "Player already in party"
+    end
+
+    if partyData.joinMode == party.JOIN_MODE_PUBLIC then
+        party.addMember( id, ply )
+        return true
+    end
+
+    if partyData.joinMode == party.JOIN_MODE_STEAM_FRIENDS and partyData.owner:IsFriendsWith( ply ) then
```
Comment (samuelWilliams99) [inline gamemode/modules/party/sv_party.lua:198]: Trigger a friend update, make attempt join run again on response

File: gamemode/modules/party/sv_party.lua [hunk 13]
```diff
@@ -0,0 +1,322 @@
...
+
+    if table.HasValue( partyData.members, ply ) then
+        return false, "Player already in party"
+    end
+
+    if #partyData.members == GAMEMODE.Config.Party.MAX_PLAYERS then
+        return false, "Party is full"
+    end
+
+    -- TODO: Add cooldown
```
Comment (samuelWilliams99) [inline gamemode/modules/party/sv_party.lua]: Put cooldown in config

File: gamemode/modules/party/sv_party.lua [hunk 14]
```diff
@@ -0,0 +1,322 @@
...
+    net.Broadcast()
+end )
+
+hook.Add( "PlayerCanSeePlayersChat", "RVR_Party_chat", function( text, isTeam, listener, speaker )
+    if not isTeam then return end
+
+    local speakerPartyID = speaker:GetPartyID()
+    if not speakerPartyID then return false end
+
+    return speakerPartyID == listener:GetPartyID()
```
Comment (samuelWilliams99) [inline gamemode/modules/party/sv_party.lua]: Change to use ply:IsInSameParty

File: gamemode/modules/party/sv_party_commands.lua [hunk 15]
```diff
@@ -0,0 +1,246 @@
...
+local function invitePlayer( caller, ply )
+    local partyData = caller:GetParty()
+
+    if not partyData then
+        return "You're not in a party"
+    end
+
+    local success, err = party.invite( partyData.id, caller, ply )
+
+    if success then
```
Comment (samuelWilliams99) [inline gamemode/modules/party/sv_party_commands.lua]: Swap this around to RERO on fail

File: gamemode/modules/party/sv_party_commands.lua [hunk 16]
```diff
@@ -0,0 +1,246 @@
...
+end )
+
+local function createParty( caller, name, tag, color, joinMode )
+    if joinMode < 0 or joinMode > 2 then
+        return "Invalid join mode, must be 0 (PUBLIC), 1 (STEAM_FRIENDS), or 2 (INVITE_ONLY)"
+    end
+
+    local success, err = party.createParty( name, caller, tag, color, joinMode )
+
+    if success then
```
Comment (samuelWilliams99) [inline gamemode/modules/party/sv_party_commands.lua]: swap for consistency

File: gamemode/modules/party/sv_party_commands.lua [hunk 17]
```diff
@@ -0,0 +1,246 @@
...
+
+        local partyData = party.getByName( str )
+        if partyData then
+            return partyData
+        end
+
+        return nil, "Unrecognised party, use party ID, name, ^ or @"
+    end )
+
+    -- TODO: Move this to commands
```
Comment (samuelWilliams99) [inline gamemode/modules/party/sv_party_commands.lua]: do that

**Review by samuelWilliams99 — COMMENTED** [review]
Results of GCR

File: gamemode/modules/party/cl_party.lua (hunk 3 above)
Comment (samuelWilliams99) [reply gamemode/modules/party/cl_party.lua]: Added a card to deal with this

File: gamemode/modules/party/sv_party.lua (hunk 12 above)
Comment (samuelWilliams99) [reply gamemode/modules/party/sv_party.lua:198]: Added a card to deal with this

File: gamemode/modules/hud/cl_hud.lua [hunk 18]
```diff
@@ -77,3 +81,38 @@ function GM:HUDShouldDraw( name )
...
+function GM:HUDDrawTargetID()
+    local trace = LocalPlayer():GetEyeTrace()
+    local aimEnt = trace.Entity
+
+    if type( aimEnt ) ~= "Player" then return end
+
+    local range = GAMEMODE.Config.HUD.TARGET_ID_RANGE
+    if LocalPlayer():GetShootPos():DistToSqr( trace.HitPos ) > range ^ 2 then return end
+
+    local text = aimEnt:Nick()
```
Comment (plally) [inline gamemode/modules/hud/cl_hud.lua]:
```suggestion
    local nickname = aimEnt:Nick()
```

File: gamemode/modules/hud/cl_hud.lua [hunk 19]
```diff
@@ -77,3 +81,38 @@ function GM:HUDShouldDraw( name )
 
     return self.BaseClass.HUDShouldDraw( self, name )
 end
+
+function GM:HUDDrawTargetID()
+    local trace = LocalPlayer():GetEyeTrace()
+    local aimEnt = trace.Entity
+
+    if type( aimEnt ) ~= "Player" then return end
```
Comment (plally) [inline gamemode/modules/hud/cl_hud.lua]:
```suggestion
    if aimEnt:IsPlayer() then return end
```

File: gamemode/modules/party/cl_party.lua [hunk 20]
```diff
@@ -0,0 +1,162 @@
+RVR.Party = RVR.Party or {}
+local party = RVR.Party
+
+function party.tryCreateParty( name, tag, color, joinMode, callback )
+    if joinMode < 0 or joinMode > 2 then return end
```
Comment (plally) [inline gamemode/modules/party/cl_party.lua]: could you use the joinMode constants here or perhaps the length of `party.joinModeStrs`

File: gamemode/modules/party/cl_party.lua [hunk 21]
```diff
@@ -0,0 +1,162 @@
...
+end
+
+hook.Add( "RVR_TargetID", "RVR_Party", function( ply, x, y )
+    local partyData = ply:GetParty()
+    if not partyData then return end
+
+    local font = "TargetID"
+    surface.SetFont( font )
+
+    local text = partyData.name
```
Comment (plally) [inline gamemode/modules/party/cl_party.lua]: partyName?

File: gamemode/modules/party/sv_party.lua [hunk 22]
```diff
@@ -0,0 +1,359 @@
...
+util.AddNetworkString( "RVR_Party_updateClient" )
+util.AddNetworkString( "RVR_Party_createParty" )
+util.AddNetworkString( "RVR_Party_updateFriends" )
+
+local function updateClientPartyData( id )
+    local partyData = party.getParty( id )
+
+    net.Start( "RVR_Party_updateClient" )
+        net.WriteUInt( id, 32 )
+        net.WriteBool( tobool( partyData ) )
```
Comment (plally) [inline gamemode/modules/party/sv_party.lua]: could you do `partyData ~= nil`?

File: gamemode/modules/party/sv_party.lua [hunk 23]
```diff
@@ -0,0 +1,359 @@
...
+end
+
+function party.setJoinMode( id, mode )
+    local partyData = party.getParty( id )
+
+    if not partyData then
+        return false, "Party with id " .. id .. " does not exist"
+    end
+
+    if mode < 0 or mode > 2 then
```
Comment (plally) [inline gamemode/modules/party/sv_party.lua]: same comment as earlier, use joinmod constants or the length of the joinmode table

File: gamemode/modules/hud/cl_hud.lua (hunk 19 above)
Comment (samuelWilliams99) [reply gamemode/modules/hud/cl_hud.lua]: That'll error if aimEnt is `nil`

File: gamemode/modules/party/cl_party.lua (hunk 20 above)
Comment (samuelWilliams99) [reply gamemode/modules/party/cl_party.lua]: I could do `if not party.joinModeStrs[joinMode] then`?

Comment (plally) [reply gamemode/modules/party/cl_party.lua]: that would be better

**Pull request merged**